File: 2º/PDOO/ProyectoRUBY/lib/vista_textual.rb
#encoding:utf-8
#require_relative 'operaciones_juego'
require 'io/console'
#require_relative "salidas_carcel.rb"
#require_relative"gestiones_inmobiliarias.rb"
#require_relative"jugador.rb"
require_relative"respuestas.rb"
module Civitas
  class Vista_textual
    def initialize
      @iGestion=0
      @iPropiedad=0
      @juegoModel
    end
    def mostrar_estado(estado)
      puts estado
    end  
    def pausa
      print "Pulsa una tecla"
      STDIN.getch
      print "\n"
    end

    def lee_entero(max,msg1,msg2)
      ok = false
      begin
        print msg1
        cadena = gets.chomp
        begin
          if (cadena =~ /\A\d+\Z/)###
            numero = cadena.to_i
            ok = true
          else
            raise IOError 
          end
        rescue IOError
          puts msg2
        end
        if (ok)
          if (numero >= max)
            ok = false
          end
        end
      end while (!ok)

      return numero
    end
    def menu(titulo,lista) 
      tab = "  "
      puts titulo
      index = 0
      lista.each { |l|
        puts tab+index.to_s+"-"+l.to_s
        index += 1
      }

      opcion = lee_entero(lista.length,
                          "\n"+tab+"Elige una opcion: ",
                          tab+"Valor erroneo")
      return opcion
    end    
    def comprar
      if(@juegoModel.getCasillaActual.instance_of?(C_calle))
        if(@juegoModel.getCasillaActual.tituloPropiedad.propietario!=nil)then 
          puts("Esta casilla no se puede comprar")
          cadena=Respuestas::NO 
          return cadena
        end
        lista=[Respuestas::NO,Respuestas::SI]
        num=menu("#{@juegoModel.getJugadorActual.nombre}, ¿Quiere comprar la propiedad #{@juegoModel.getCasillaActual.nombre}?", lista)
        cadena=lista[num]
      else
        puts("Esta casilla no se puede comprar")
        cadena=Respuestas::NO
      end
      return cadena
    end
    def gestionar
        cadena="Que tipo de gestion inmobiliaria desearealizar: "
        lista1=[GestionesInmobiliarias::VENDER, GestionesInmobiliarias::HIPOTECAR, GestionesInmobiliarias::CANCELAR_HIPOTECA, GestionesInmobiliarias::CONSTRUIR_CASA, GestionesInmobiliarias::CONSTRUIR_HOTEL, GestionesInmobiliarias::TERMINAR]
        @iGestion=lista1[menu(cadena,lista1)]
        lista2=[]
        cadena="Que propiedad quiere gestionar: "
        if(@iGestion==GestionesInmobiliarias::CANCELAR_HIPOTECA)then
          propiedades=@juegoModel.getJugadorActual().propiedades
          propiedades.each do |propiedad|
            if(propiedad.hipotecado==true)then
              lista2<<propiedad
            end
          end
          if(lista2.size==0)
            @iPropiedad=-1
          else
            @iPropiedad=menu(cadena, lista2)
          end
        elsif(@iGestion==GestionesInmobiliarias::HIPOTECAR) 
          propiedades=@juegoModel.getJugadorActual().propiedades
          propiedades.each do |propiedad|
            if(propiedad.hipotecado==false)then
              lista2<<propiedad
            end
          end
          @iPropiedad=menu(cadena, lista2)
        elsif(@iGestion==GestionesInmobiliarias::VENDER)
          propiedades=@juegoModel.getJugadorActual().propiedades
          propiedades.each do |propiedad|
              lista2<<propiedad
          end
          @iPropiedad=menu(cadena, lista2)
        elsif(@iGestion==GestionesInmobiliarias::CONSTRUIR_CASA)
          propiedades=@juegoModel.getJugadorActual().propiedades
          propiedades.each do |propiedad|
            lista2<<propiedad
          end
          @iPropiedad=menu(cadena, lista2)        
        elsif(@iGestion==GestionesInmobiliarias::CONSTRUIR_HOTEL)
          propiedades=@juegoModel.getJugadorActual().propiedades
          propiedades.each do |propiedad|
              lista2<<propiedad
          end
          @iPropiedad=menu(cadena, lista2)   
        elsif(@iGestion==GestionesInmobiliarias::TERMINAR)then
           @iPropiedad=-1
        end 
    end
    def getGestion
      return @iGestion
    end
    def getPropiedad
      return @iPropiedad
    end

    def mostrarSiguienteOperacion(operacion)
      puts("La siguiente operacion es #{operacion}")
    end

    def mostrarEventos
      while(Diario.instance.eventos_pendientes==true) 
        puts(Diario.instance.leer_evento)
      end
    end

    def setCivitasJuego(civitas)
         @juegoModel=civitas
         self.actualizarVista
    end

    def actualizarVista
      puts("----------------------Actualizar vista----------------------------")
      jugador=@juegoModel.getJugadorActual
      casilla=@juegoModel.getCasillaActual
      puts(jugador.to_s)
      puts(casilla.to_s)
      puts("------------------------------------------------------------------")
      #muestra informacion del jugador actual, sus propiedades y la casilla
    end
    def salirCarcel()
      lista=[SalidasCarcel::TIRANDO, SalidasCarcel::PAGANDO]
      num=menu("¿Como prefiere salir de la carcel?",lista)
      cadena=lista[num]
      return cadena
    end
    
    
  end

end

File: 2º/PDOO/ProyectoJAVA/src/JuegoTexto/VistaTextual.java
package JuegoTexto;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;
import civitas.CivitasJuego;
import civitas.Diario;
import civitas.OperacionesJuego;
import civitas.Respuestas;
import civitas.SalidasCarcel;

class VistaTextual {
  
  CivitasJuego juegoModel; 
  int iGestion=-1;
  int iPropiedad=-1;
  private static String separador = "=====================";
  
  private Scanner in;
  
  VistaTextual () {
    in = new Scanner (System.in);
  }
  
  void mostrarEstado(String estado) {
    System.out.println (estado);
  }
              
  void pausa() {
    System.out.print ("Pulsa una tecla");
    in.nextLine();
  }

  int leeEntero (int max, String msg1, String msg2) {
    Boolean ok;
    String cadena;
    int numero = -1;
    do {
      System.out.print (msg1);
      cadena = in.nextLine();
      try {  
        numero = Integer.parseInt(cadena);
        ok = true;
      } catch (NumberFormatException e) { // No se ha introducido un entero
        System.out.println (msg2);
        ok = false;  
      }
      if (ok && (numero < 0 || numero >= max)) {
        System.out.println (msg2);
        ok = false;
      }
    } while (!ok);

    return numero;
  }

  int menu (String titulo, ArrayList<String> lista) {
    String tab = "  ";
    int opcion;
    System.out.println (titulo);
    for (int i = 0; i < lista.size(); i++) {
      System.out.println (tab+i+"-"+lista.get(i));
    }

    opcion = leeEntero(lista.size(),
                          "\n"+tab+"Elige una opción: ",
                          tab+"Valor erróneo");
    return opcion;
  }

  SalidasCarcel salirCarcel() {
    int opcion = menu ("Elige la forma para intentar salir de la carcel",
      new ArrayList<> (Arrays.asList("Pagando","Tirando el dado")));
    return (SalidasCarcel.values()[opcion]);
  }

  Respuestas comprar(){
      String e = juegoModel.getCasillaActual().toString();
      int opcion = menu ("Elige si quieres comprar la calle "+e,
      new ArrayList<> (Arrays.asList("No comprar","Comprar")));
    return (Respuestas.values()[opcion]);
  }

  void gestionar (){
      iGestion = -1;
      while(iGestion != 5){
        ArrayList<String> propiedadesString  = new ArrayList<>();
        juegoModel.getJugadorActual().getPropiedades().forEach((p) -> {
        propiedadesString.add(p.toString());
        });
        iGestion = menu ("Elige el numero de gestion inmobiliaria",
        new ArrayList<> (Arrays.asList("Vender", "Hipotecar", "Cancelar hipoteca", "Construir casa", "Construir hotel", "Terminar")));
        if(iGestion != 5)
            iPropiedad = menu ("Elige el numero de propiedad", propiedadesString);
        
        if(iGestion ==0)
          juegoModel.vender(iPropiedad);
        else if(iGestion ==1)
          juegoModel.hipotecar(iPropiedad);
        else if(iGestion ==2)
          juegoModel.cancelarHipoteca(iPropiedad);
        else if(iGestion ==3)
          juegoModel.construirCasa(iPropiedad);
        else if(iGestion ==4)
          juegoModel.construirHotel(iPropiedad);
      }
  }
  
  public int getGestion(){
      return iGestion;
  }
  
  public int getPropiedad(){
      return iPropiedad;
  }
    

  void mostrarSiguienteOperacion(OperacionesJuego operacion) {
      System.out.println (operacion);
  }


  void mostrarEventos(){
      while(Diario.getInstance().eventosPendientes()){
          System.out.println(Diario.getInstance().leerEvento()); 
      }
  }
  
  public void setCivitasJuego(CivitasJuego civitas){ 
        juegoModel=civitas;
        //this.actualizarVista();
    }
  
  void actualizarVista(){
    mostrarEstado("Vista:\n>"+juegoModel.getJugadorActual().toString() + juegoModel.getCasillaActual().toString());
  } 
}
